### PR TITLE
GH-104142: Fix `_Py_RefcntAdd` to respect immortality

### PR DESCRIPTION
**Pull request by brandtbucher**
Also, make the existing immortality tests a bit stricter and more comprehensive.

* Issue: gh-104142

### CODE REVIEW

**Comment by ericsnowcurrently** [comment]
CC @eduardo-elizondo

**Comment by brandtbucher** [comment]
(Not sure why buildbots aren't running...)

**Review by corona10 — DISMISSED** [review]

**Comment by corona10** [comment]
Ah would you like to add the C API test in here too?
https://github.com/python/cpython/blob/9885677b0494e9be3eb8d7d69bebca0e79d8abcc/Modules/_testcapi/immortal.c#L34-L38

**Comment by corona10** [comment]
Ah ignore https://github.com/python/cpython/pull/104143#issuecomment-1533983946
it's not public API...

**Review by corona10 — APPROVED** [review]

File: Include/internal/pycore_object.h [hunk 1]
```diff
@@ -58,6 +58,9 @@ extern void _Py_DecRefTotal(PyInterpreterState *);
 // Increment reference count by n
 static inline void _Py_RefcntAdd(PyObject* op, Py_ssize_t n)
 {
+    if (_Py_IsImmortal(op)) {
+        return;
+    }
```
Comment (eduardo-elizondo) [inline Include/internal/pycore_object.h:63]: One quick comment here is that if this were to be in a hot path, I would recommend using the same technique that we use for `Py_INCREF` to minimize the number of generated instructions to do the check and add: https://github.com/python/cpython/blob/main/Include/object.h#L620-L634

Now, given that it's only in use for `sq_repeat` in tuple and list, I don't think this is perf sensitive so keeping the code as you have it here should be good.

File: Lib/test/test_builtin.py [hunk 2]
```diff
@@ -2372,24 +2372,31 @@ def __del__(self):
...
+        IMMORTAL_REFCOUNT = (1 << 32) - 1
+
+    IMMORTALS = (None, True, False, Ellipsis, NotImplemented, *range(-5, 257))
+
+    def assert_immortal(self, immortal):
+        with self.subTest(immortal):
+            self.assertEqual(sys.getrefcount(immortal), self.IMMORTAL_REFCOUNT)
+
+    def test_immortals(self):
+        for immortal in self.IMMORTALS:
```
Comment (eduardo-elizondo) [inline Lib/test/test_builtin.py:2388]: Small nit: I generally prefer rolling out assertions in unit tests to make it easier to isolate by line-number in the place where the test failed. But I don't think there's a standard in the code base so I'm fine either way

**Comment by eduardo-elizondo** [comment]
Nice catch, I could have sworn that I had added the check to `_Py_RefcntAdd ` but I somehow missed it. I added two small notes but they are minor so this looks good to me

Thanks for the fix!

**Review by eduardo-elizondo — APPROVED** [review]

**Comment by sunmy2019** [comment]
LGTM, all increments to `ob->ob_refcnt` should be guarded now.

File: Lib/test/test_builtin.py (hunk 2 above)
Comment (brandtbucher) [reply Lib/test/test_builtin.py:2388]: Yeah, I'd like comprehensive testing of these immortals though, and I don't want to roll out ~250 different assertions for each test.

I figured making each one a `subTest` in `assert_immortal` is a pretty good compromise.

**Comment by JelleZijlstra** [comment]
There is a USan failure:

```
Include/object.h:227:12: runtime error: member access within address 0x7fff7ae78b80 with insufficient space for an object of type 'PyObject' (aka 'struct _object')
0x7fff7ae78b80: note: pointer points here
 b8 7f 00 00  00 47 a8 c2 97 55 00 00  00 b5 59 7e 78 cd 65 f5  00 8a ea c2 b8 7f 00 00  39 7a 90 c0
              ^ 
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior Include/object.h:227:12 in 
make: *** [Makefile:1106: checksharedmods] Error 1
```

Not sure what that means, but the line is in _Py_IsImmortal (https://github.com/brandtbucher/cpython/blob/py-refcnt-add-immortal/Include/object.h#L227), so it may be real.

**Comment by JelleZijlstra** [comment]
(Trying again just in case it was something flaky.)

**Comment by brandtbucher** [comment]
Grrr... it failed again.

**Comment by brandtbucher** [comment]
I honestly don't know what to do here.

It *feels* like a bug in USan, since I honestly don't get how the new code could cause that failure. But it's reproducible, and indeed seems to indeed be related to my change here.

**Comment by brandtbucher** [comment]
Ah, wait, the same buildbot is also failing on `main`: https://buildbot.python.org/all/#/builders/719/builds/2611

So it's an existing issue (but one that probably should still be looked into).

**Comment by sunmy2019** [comment]
> So it's an existing issue (but one that probably should still be looked into).

See #104190

**Comment by brandtbucher** [comment]
Nice catch!

**Pull request merged**